app/gui2/shared/languageServer.ts

```diff
@@ -91,57 +98,131 @@ export class RemoteRpcError {
   }
 }
 
-export class LsRpcError extends Error {
-  cause: RemoteRpcError | Error
+export class LsRpcError {
+  cause: RemoteRpcError | Error | string
   request: string
   params: object
-  constructor(cause: RemoteRpcError | Error, request: string, params: object) {
-    super(`Language server request '${request}' failed.`)
+  constructor(cause: RemoteRpcError | Error | string, request: string, params: object) {
     this.cause = cause
     this.request = request
     this.params = params
   }
+
+  toString() {
+    return `Language server request '${this.request} failed: ${this.cause instanceof RemoteRpcError ? this.cause.message : this.cause}`
+  }
 }
 
-/** [Documentation](https://github.com/enso-org/enso/blob/develop/docs/language-server/protocol-language-server.md) */
+export type LsRpcResult<T> = Result<T, LsRpcError>
+
+/**
+ * This client implements the [Language Server Protocol](https://github.com/enso-org/enso/blob/develop/docs/language-server/protocol-language-server.md)
+ *
+ * It also handles the initialization (and re-initialization on every reconnect); each method
+ * repressenting a remote call (except the `initProtocolConnection` obviously) waits for
+ * initialization before sending the request.
+ */
 export class LanguageServer extends ObservableV2<Notifications> {
   client: Client
-  handlers: Map<string, Set<(...params: any[]) => void>>
-  retainCount = 1
-
-  constructor(client: Client) {
+  /**
+   * This promise is resolved once the LS protocol is initialized. When connection is lost, this
+   * field becomes again an unresolved promise until reconnected and reinitialized.
+   */
+  initialized: Promise<LsRpcResult<response.InitProtocolConnection>>
+  private clientScope: AbortScope = new AbortScope()
+  private initializationScheduled = false
+  private retainCount = 1
+
+  constructor(
+    private clientID: Uuid,
+    private transport: ReconnectingTransportWithWebsocketEvents,
+  ) {
     super()
-    this.client = client
-    this.handlers = new Map()
-
-    client.onNotification((notification) => {
+    this.initialized = this.scheduleInitializationAfterConnect()
+    const requestManager = new RequestManager([transport])
+    this.client = new Client(requestManager)
+    this.client.onNotification((notification) => {
       this.emit(notification.method as keyof Notifications, [notification.params])
     })
-    client.onError((error) => {
+    this.client.onError((error) => {
       console.error(`Unexpected LS connection error:`, error)
     })
+    transport.on('error', (error) => console.error('Language Server transport error:', error))
+    const reinitializeCb = () => {
+      console.log('Language Server: websocket closed')
+      this.scheduleInitializationAfterConnect()
+    }
+    transport.on('close', reinitializeCb)
+    this.clientScope.onAbort(() => {
+      this.transport.off('close', reinitializeCb)
+      this.transport.close()
+    })
+  }
+
+  private scheduleInitializationAfterConnect() {
```

Comment:
And this is valid: this function does not connect, because it does not make the reconnect (the websocket underneath does), only listens for "open" event to start initialization after.